### ISSUE TITLE
Multiple services and multi modules.

### ISSUE THREAD

**Issue by itscorey** [apollographql/apollo-android#3033]
**Question.**

**TLDR**
> Multiple services with multi modules doesn't work well if you define a new service in the common module with no sources (no fragments) and wants to use it in a feature module. It'll work just fine if you define a dumb Fragment and the sources will be generated in the feature module.

----

Hi folks,

I've been using the multi modules feature and it have been great by now. In my setup I've the following configuration:

A `shared` module containing common `Fragments` and `Types`. But this module also defines the following services:

```groovy
// shared build.gradle
apollo {
   generateApolloMetadata.set(true)
   
    // there is also rootPackageName and sourceFolder
    // because each service will have its own source folder like
    // com.myapp.shared.apione/fragments/*
       
    service("apiOne") {
        schemaPath = "../../schemas/apiOne.json"
    }
    
    service("apiTwo") {
        schemaPath = "../../schemas/apiTwo.json"
    }
    
    customTypeMapping = [] // ...
}
```

Now each feature module is able to depends on this shared module and get access to the generated `Fragment` when necessary. Each feature module is configured like so:

```groovy
apollo {
    // schema is only defined in shared module
    
    // each feature defines services they use so they can define their own queries/fragments
    service("apiOne") {
        rootPackageName = "${featurePackage}.api.apione" // com.myapp.dashboard
        sourceFolder = "com/myapp/api/apione/"
    }
    
    service("apiTwo") {
        rootPackageName = "${featurePackage}.api.apitwo" // com.myapp.dashboard
        sourceFolder = "com/myapp/api/apitwo/"
    }
}
```

They also depends on `project(":shared")` and uses `apolloMetadata` from it.

By doing this we have the following advantages:

 - We can share common fragments and types across multiple modules
 - We can also shared our schemas files which means that our feature modules don't need to have it anymore.
 
 **First question**: is this configuration right to integra multiple services with multi modules? The documentation does not says much about it and in the second question below I explain a problem we're having with it.
 
 **Second question**: we introduced a new service and something bad happened: we aren't able to generate sources of the feature modules because apollo couldn't provide enough information about the new service and this is... strange.
 
 This happened because our `shared` module didn't have any GQL files inside the source folder of the new service. If we put a dumb `Fragment` or `Query` inside the source folder of the new service we are able to generate the sources for the features modules.
 
 Thank you all in advance.
 
 Best,
 Lucas.

**Comment by itscorey** [comment]
The dependency graph is the following:

![image](https://user-images.githubusercontent.com/20311112/113752721-ab437700-96e3-11eb-9d07-e61137bb2e0c.png)

**Comment by martinbonnin** [comment]
Ah, good catch! And thanks for the very detailed bug report. I made a reproducer there: https://github.com/martinbonnin/apollo-android-samples/tree/main/multi-modules-multi-services

Indeed there's something wrong with the Gradle config so that when there's no graphql file, the task will be completely skipped with `NO-SOURCE` and will not write the metadata. As you found out, the workaround is to add an empty `.graphql` file in the `shared` module. The longer term fix is to refactor a lot of the Gradle logic (see also https://github.com/apollographql/apollo-android/pull/2668)

**Comment by itscorey** [comment]
> Ah, good catch! And thanks for the very detailed bug report. I made a reproducer there: https://github.com/martinbonnin/apollo-android-samples/tree/main/multi-modules-multi-services
> 
> Indeed there's something wrong with the Gradle config so that when there's no graphql file, the task will be completely skipped with `NO-SOURCE` and will not write the metadata. As you found out, the workaround is to add an empty `.graphql` file in the `shared` module. The longer term fix is to refactor a lot of the Gradle logic (see also #2668)

I see! Thank you for your response, Martin. The short-term fix is fine and I'll go with it.

This indeed solved my problem. I thought I did something wrong while setting up the shared module.

Thank you again!
